### PR TITLE
Feature/blocking users

### PR DESCRIPTION
**Pull request by nerdsupremacist**

### CODE REVIEW

File: src/main/java/backend/controller/AdminController.kt [hunk 1]
```diff
@@ -92,6 +92,6 @@ class AdminController(private val mailService: MailService,
         val team = teamService.findOne(body.teamId) ?: throw NotFoundException("Team with ID ${body.teamId} not found")
         return PostingView(
                 postingService.adminCreatePosting(
-                        team.members.elementAt(0), "Current Location: ${body.city}", null, Coord(body.latitude, body.longitude), LocalDateTime.now()), null)
+                        team.members.elementAt(0), "Current Location: ${body.city}", null, Coord(body.latitude, body.longitude), LocalDateTime.now()), null, null)
```
Comment (thepiwo) [inline src/main/java/backend/controller/AdminController.kt]: should we consider defaulting to null for the last 2 params?

Comment (florianschmidt1994) [inline src/main/java/backend/controller/AdminController.kt]: Extract variable here for the posting, so we have less nesting

File: src/main/java/backend/controller/PostingController.kt [hunk 2]
```diff
@@ -67,20 +67,26 @@ class PostingController(private val postingService: PostingService,
         val posting = postingService.createPosting(user, body.text, body.uploadMediaTypes, locationCoord, clientDate)
         posting.media.forEach { it.uploadToken = getSignedJwtToken(JWT_SECRET, it.id.toString()) }
 
-        return PostingView(posting, null)
+        return PostingView(posting, null, user.account.id)
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/PostingController.kt]: In the long run I think we should think about whether or not we want to keep using this pattern or do something cleaner here

File: src/main/java/backend/model/UserGenerated.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        return getUser()?.isBlockedBy(userId) ?: false
+    }
+
+    override fun isBlocking(user: User?): Boolean {
+        return getUser()?.isBlocking(user) ?: false
+    }
+}
+
+fun <T: Blockable> Iterable<T>.removeBlockedBy(userId: Long?): Iterable<T> {
+    return this.filter { !it.isBlockedBy(userId) }
```
Comment (thepiwo) [inline src/main/java/backend/model/UserGenerated.kt:26]: maybe turn around the semantic meaning to notBlockedBy as it is always used in negation or false check

**Review by thepiwo — COMMENTED** [review]
I think there should be an endpoint added to see all blocked users for the logged in one, to make a UI for unblocking

File: src/main/java/backend/controller/PostingController.kt [hunk 4]
```diff
@@ -67,20 +67,26 @@ class PostingController(private val postingService: PostingService,
...
     }
 
     /**
      * GET /posting/{id}/
      * Gets posting by id
      */
     @GetMapping("/{id}/")
-    fun getPosting(@PathVariable("id") id: Long, @RequestParam(value = "userid", required = false) userId: Long?): PostingResponseView {
+    fun getPosting(@PathVariable("id") id: Long,
+                   @RequestParam(value = "userid", required = false) userId: Long?): PostingResponseView {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/PostingController.kt]: This should not be userid as request param, but can be fetched from spring context of the currently authenticated user.

Maybe try out @AuthenticationPrincipal and see what happens

File: src/main/java/backend/controller/PostingController.kt [hunk 5]
```diff
@@ -210,10 +216,12 @@ class PostingController(private val postingService: PostingService,
      * Gets Likes for Posting
      */
     @GetMapping("/{id}/like/")
-    fun getLikesForPosting(@PathVariable("id") id: Long): List<LikeView> {
+    fun getLikesForPosting(@PathVariable("id") id: Long,
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/PostingController.kt]: See above

File: src/main/java/backend/controller/TeamController.kt [hunk 6]
```diff
@@ -211,8 +212,13 @@ class TeamController(private val teamService: TeamService,
      * gets a specific Team
      */
     @GetMapping("/{teamId}/")
-    fun showTeam(@PathVariable teamId: Long): TeamView {
+    fun showTeam(@PathVariable teamId: Long,
+                 @RequestParam(value = "userid", required = false) userId: Long?): TeamView {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/TeamController.kt]: See above

File: src/main/java/backend/controller/TeamController.kt [hunk 7]
```diff
@@ -224,10 +230,12 @@ class TeamController(private val teamService: TeamService,
      */
     @Cacheable(TEAMS, sync = true)
     @GetMapping("/")
-    fun showTeamsByEvent(@PathVariable eventId: Long): Iterable<TeamView> {
+    fun showTeamsByEvent(@PathVariable eventId: Long,
+                         @RequestParam(value = "userid", required = false) userId: Long?): Iterable<TeamView> {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/TeamController.kt]: See above

File: src/main/java/backend/controller/UserController.kt [hunk 8]
```diff
@@ -101,20 +105,23 @@ class UserController(private val userService: UserService,
      * Gets all users
      */
     @GetMapping("/")
-    fun showUsers(): Iterable<BasicUserView> {
-        return userService.getAllUsers().map(::BasicUserView)
+    fun showUsers(@RequestParam(value = "userid", required = false) userId: Long?): Iterable<BasicUserView> {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/UserController.kt]: See above

File: src/main/java/backend/controller/UserController.kt [hunk 9]
```diff
@@ -101,20 +105,23 @@ class UserController(private val userService: UserService,
...
     }
 
     /**
      * GET /user/search/{search}/
      * Searches for User by String greater 2 chars
      */
     @GetMapping("/search/{search}/")
-    fun searchUsers(@PathVariable("search") search: String): List<SimpleUserView> {
+    fun searchUsers(@PathVariable("search") search: String,
+                    @RequestParam(value = "userid", required = false) userId: Long?): List<SimpleUserView> {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/UserController.kt]: See above

File: src/main/java/backend/controller/UserController.kt [hunk 10]
```diff
@@ -170,9 +177,60 @@ class UserController(private val userService: UserService,
      * Gets user with given id
      */
     @GetMapping("/{id}/")
-    fun showUser(@PathVariable id: Long): BasicUserView {
+    fun showUser(@PathVariable id: Long,
+                 @RequestParam(value = "userid", required = false) userId: Long?): BasicUserView {
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/UserController.kt]: See above

**Comment by florianschmidt1994** [comment]
Could you also add some tests for this that verify that blocking, unblocking and the corresponding returned entities in the view are hidden as expected pls!

Thanks! Good work

File: src/main/java/backend/controller/UserController.kt [hunk 11]
```diff
@@ -196,13 +196,15 @@ class UserController(private val userService: UserService,
...
     @PreAuthorize("isAuthenticated()")
-    @PostMapping("/{id}/block")
+    @PostMapping("/{id}/block/")
     fun blockUser(@PathVariable id: Long,
                   @AuthenticationPrincipal customUserDetails: CustomUserDetails): BasicUserView {
 
         val currentUser = userService.getUserFromCustomUserDetails(customUserDetails)
         val user = userService.getUserById(id) ?: throw NotFoundException("user with id $id does not exist")
 
+        if (user.account.id == currentUser.account.id) throw BadRequestException("user not allowed to block itself")
```
Comment (florianschmidt1994) [inline src/main/java/backend/controller/UserController.kt]: Nice!

**Review by thepiwo — APPROVED** [review]

**Pull request merged**